pycontrails/datalib/ecmwf/model_levels.py

```diff
@@ -1,79 +1,412 @@
+        may consume a large amount of memory.
+
+        The ``dtype`` of the output is the same as the ``dtype`` of the ``sp`` parameter.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> import xarray as xr
+
+    >>> sp_arr = np.linspace(101325.0, 90000.0, 16).reshape(4, 4)
+    >>> longitude = np.linspace(-180, 180, 4)
+    >>> latitude = np.linspace(-90, 90, 4)
+    >>> sp = xr.DataArray(sp_arr, coords={"longitude": longitude, "latitude": latitude})
+
+    >>> model_levels = [80, 100]
+    >>> pressure_level_at_model_levels(sp, model_levels)
+    <xarray.DataArray (model_level: 2, longitude: 4, latitude: 4)> Size: 256B
+    array([[[259.75493944, 259.27107504, 258.78721064, 258.30334624],
+            [257.81948184, 257.33561744, 256.85175304, 256.36788864],
+            [255.88402424, 255.40015984, 254.91629544, 254.43243104],
+            [253.94856664, 253.46470224, 252.98083784, 252.49697344]],
+           [[589.67975444, 586.47283154, 583.26590864, 580.05898574],
+            [576.85206284, 573.64513994, 570.43821704, 567.23129414],
+            [564.02437124, 560.81744834, 557.61052544, 554.40360254],
+            [551.19667964, 547.98975674, 544.78283384, 541.57591094]]])
+    Coordinates:
+      * longitude    (longitude) float64 32B -180.0 -60.0 60.0 180.0
+      * latitude     (latitude) float64 32B -90.0 -30.0 30.0 90.0
+      * model_level  (model_level) int64 16B 80 100
+
+    See Also
+    --------
+    pressure_levels_at_model_levels_constant_surface_pressure
+    """
+    model_levels = np.asarray(model_levels, dtype=int)
+    if not np.all((model_levels >= 1) & (model_levels <= 137)):
+        msg = "model_levels must be integers between 1 and 137"
+        raise ValueError(msg)
+
+    usecols = ["n", "a [Pa]", "b"]
+    df = (
+        pd.read_csv(MODEL_LEVELS_PATH, usecols=usecols)
+        .rename(columns={"n": "model_level", "a [Pa]": "a"})
+        .set_index("model_level")
+    )
+
+    a = df["a"].to_xarray()
+    b = df["b"].to_xarray()
+
+    if "model_level" in sp.dims:
+        sp_model_levels = sp["model_level"]
+        if len(sp_model_levels) != 1:
+            msg = "Found multiple model levels in sp, expected at most one"
+            raise ValueError(msg)
+        if sp_model_levels.item() != 1:
+            msg = f"sp must be at model level 1, found model level {sp_model_levels.item()}"
+            raise ValueError(msg)
+        # Remove the model level dimension to allow automatic broadcasting below
+        sp = sp.squeeze("model_level")
+
+    dtype = sp.dtype
+    a = a.astype(dtype, copy=False)
+    b = b.astype(dtype, copy=False)
+
+    indexer = {"model_level": model_levels}
+    p_half_below = a.sel(indexer) + b.sel(indexer) * sp
+
+    indexer = {"model_level": model_levels - 1}
+    p_half_above = (a.sel(indexer) + b.sel(indexer) * sp).assign_coords(model_level=model_levels)
+
+    p_full = (p_half_above + p_half_below) / 2.0
+    return p_full / 100.0  # Pa -> hPa
```

Comment:
I think that this could happen when the surface pressure is below around 30,000 Pa. I can check this in the implementation and issue a warning in this case.

Reply:
1f7bc874a4485d5f39e8ac71660e279140211663